pkg/clustermesh/common/sync.go

```diff
@@ -0,0 +1,99 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package common
+
+import (
+	"context"
+
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/spanstat"
+)
+
+// SyncState is used to track the synchronization state of clusters to the local kvstore.
+// For clustermesh-apiserver, this will be synchronization of the local cluster.
+// For kvstoremesh, this will be synchronization of remote clusters.
+type SyncState struct {
+	lock.StoppableWaitGroup
+	Init          spanstat.SpanStat
+	CompletedChan chan struct{}
+}
+
+// NewSyncState creates a new SyncState for tracking the progress of
+// synchronization to the local kvstore.
+func NewSyncState(apimetrics ApiserverMetrics) *SyncState {
+	ss := &SyncState{
+		StoppableWaitGroup: *lock.NewStoppableWaitGroup(),
+		CompletedChan:      make(chan struct{}),
+	}
+	ss.Init.Start()
+
+	// When all resources have been synchronized, update metrics and signal to
+	// external processes by closing the CompletedChan.
+	go func() {
+		<-ss.WaitChannel()
+		ss.Init.End(true)
+		close(ss.CompletedChan)
+		apimetrics.InitialSyncDuration.Set(ss.Init.Seconds())
+	}()
+
+	return ss
+}
+
+// NewNoopSyncState creates a new SyncState that is already completed. This is
+// used as a placeholder in shared logic for processes that do not need to
+// track synchronization.
+func NewNoopSyncState() *SyncState {
+	noopSS := &SyncState{
+		StoppableWaitGroup: *lock.NewStoppableWaitGroup(),
+	}
+	noopSS.Stop()
+	return noopSS
+}
+
+// WaitForCluster adds a cluster to the SyncState and returns a callback
+// function that should be called when all of the cluster's resources been
+// synchronized.
+func (ss *SyncState) AddCluster(ctx context.Context) *ClusterSyncState {
+	ctx, cancel := context.WithCancel(ctx)
+	css := &ClusterSyncState{
+		StoppableWaitGroup: *lock.NewStoppableWaitGroup(),
+		CancelWait:         cancel,
+	}
+	ss.Add()
+	go func() {
+		// Wait for the context to be canceled or for the SyncState to be completed.
+		select {
+		case <-ctx.Done():
+		case <-css.WaitChannel():
+		}
+		ss.Done()
+	}()
+	return css
+}
+
+// Complete returns true if all clusters have been synchronized to the kvstore.
+func (ss *SyncState) Complete() bool {
+	select {
+	case <-ss.WaitChannel():
+		return true
+	default:
+		return false
+	}
+}
+
+// ClusterSyncState is used to track the synchronization of a single cluster to
+// the local kvstore.
+type ClusterSyncState struct {
```

Comment:
Does this type need to be exported?

Reply:
Nope, just its methods. I'll update that :+1: